[Dynamo] Implement a simple mutation tracker for user defined triton kernels

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116466

This PR adds a very simple mutation tracking mechanism to dynamo which can later be improved to be more thorough. Currently it allows tensors to be in tl.load but if it sees a tensor used anywhere else (including a tl.load), it bails out.

One question about the method: is `ast.NodeVisitor` the best thing to use here? Having to detect mutations with this is kinda pretty since you need to keep setting state at each transition.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

oulgen committed:
[Dynamo] Implement a simple mutation tracker for user defined triton kernels

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
+        if (
+            isinstance(node.func, ast.Attribute)
+            and isinstance(node.func.value, ast.Name)
+            and node.func.value.id == "tl"
```
What if `triton.language` is imported not as `tl`? Can we make this more robust by checking the actual module behind it? Most of the time it's `tl`, but I feel non-`tl` is a feasible case.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
+            if node.func.attr == "store":
+                assert self.in_store is False
+                self.in_store = True
+                self.generic_visit(node)
```
IIUC, `self.generic_visit` visits *all* children of the `tl.store` node from here. I think, we should be more specific than just saying "we're in store and every input covered here is mutated". This is because `tl.store` has [multiple parameters](https://triton-lang.org/main/python-api/generated/triton.language.store.html#triton.language.store) and we only care about the inputs covered by the first one: `pointer`. In other words, if any inputs are used only in the `value` and / or `mask` computation, those are (generally) not mutated.

It may be that the code path from those additional argument back to the non-modified inputs always goes through `tl.load`, though, in which case `if self.in_load: pass` will happen above. But not quite sure about this.

oulgen replied:
Keyword args and regular args get parsed differently, so doing it only for first arg was annoying. I doubt anyone will use input tensor in other arguments load so this was simple enough?

aakhundov replied:
So the current code does analysis on all args of `tl.store`, both keyword and regular? Using read-only input tensor in the dependency tree of other arguments is possible. E.g., the values loaded from read-only arg are used in the (second) `value` argument of `tl.store`. But, as far as I can think, there will always be a `tl.load` somewhere on the code path from the other (not `pointer`) arguments to the read-only inputs. So, if you think it is non-trivial to consider only the first argument of `tl.store`, it sounds fine to start with this version and see if we run into edge cases later.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
+        elif self.in_store:
+            self.infos[node.id].mutated = True
+        else:
+            self.infos[node.id].used_in_unknown = True
```
IIUC, this means if the parameters are used in any call that is not `tl.load`, we're going to claim they are mutated. But then it looks a bit too restrictive to me. Can we extend the current no-mutation behavior of `tl.load` to `tl.*` except `tl.store` (because, as far as I know, from all `tl.*`, only `tl.store` actually writes into the pointer)? E.g., if I add `tl.device_print(x)` to the kernel code, the current version of the PR will declare `x` mutated, right?

oulgen replied:
I’d rather have allow list of what loads rather than assume everything except for store does not mutate. But this can just be added on case by case basis

aakhundov replied:
Do you know about any other `tl.*` function that mutates the data behind the tensor pointer argument(s)? Quickly skimming through [the list here](https://triton-lang.org/main/python-api/triton.language.html) makes me believe it's only `tl.store`. And then there are (at the bottom of the list) compiler hint ops and debug ops that could be called with pointer args and will lead to unnecessary mutation flagging with the current code?

I totally agree than all non-`tl` calls should be marked as suspicious. But then I'm wondering how deep does our static analysis go: is it just inside the main Triton kernel function's body or also the (potentially nested) helper functions called from inside it? Do we also dive inside the `tl.*` functions?

aakhundov replied:
If you're inclined to have an allow list, let's add one with the compiler hint and debug ops in it? I'll be happy with that at this stage, as long as we don't go inside the `tl.*` functions (and treat helper functions as unknown).

aakhundov replied:
For posterity: we've discussed this with @oulgen offline. The AST visitor analysis in this PR considers every statement (in the Triton kernel code) in isolation, hence doesn't go into the body of the `tl.*` and other function calls.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
+
+    infos = {name: MutationInfo() for name in tensors}
+    tracker = MutationTracker(infos)
+    tracker.visit(kernel.parse())
```
is `kernel.parse()` doing the same thing as `ast.parse(kernel)`?

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -54,6 +56,70 @@ def reset_table(self) -> None:
+                self.in_store = False
+                return
+            if node.func.attr == "load":
+                assert self.in_load is False
```
I think, in indirect indexing, we can run into nested `tl.load`s? This will break then.

oulgen replied:
Can you give an example?

oulgen replied:
This would require us to restore state if there are nested stuff

aakhundov replied:
For example, smth like (similar to what exists in many jagged tensor Triton kernels):

```
index = tl.load(index_ptr + offsets)
data = tl.load(data_ptr + index, mask)
```

Maybe we could use a counter instead of a boolean flag for tracking inside nested `loads`?

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -179,7 +249,7 @@ def triton_kernel_wrapper_functional_dense(
     }
     triton_kernel_wrapper_mutation(kernel_idx=kernel_idx, grid=grid, kwargs=kwargs)
-    return kwargs
+    return {key: val for key, val in kwargs.items() if key in tensors_to_clone}
```
Why did we return all kwargs before and now only cloned tensor kwargs? Is it because now we effectively don't know the real outputs and don't care about returning the rest?

oulgen replied:
Previously everything was cloned but not some arguments are input only. This is required for autograd to work correctly

aakhundov reviewed (changes requested):
Thanks @oulgen! Left a few comments.

oulgen committed:
Update on "[Dynamo] Implement a simple mutation tracker for user defined triton kernels"


This PR adds a very simple mutation tracking mechanism to dynamo which can later be improved to be more thorough. Currently it allows tensors to be in tl.load but if it sees a tensor used anywhere else (including a tl.load), it bails out.

One question about the method: is `ast.NodeVisitor` the best thing to use here? Having to detect mutations with this is kinda pretty since you need to keep setting state at each transition.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -54,6 +56,95 @@ def reset_table(self) -> None:
+                self.generic_visit(node)
+                self.read_depth -= 1
+                return
+        self.generic_visit(node)
```
Missed this one before: do we need to do `self.generic_visit(node)` again after the one in one of the `if` statements above? If not, should we make the whole thing `if ... else if ... else`?

aakhundov replied:
Scratch this: there are `return`s.

oulgen committed:
Update on "[Dynamo] Implement a simple mutation tracker for user defined triton kernels"


This PR adds a very simple mutation tracking mechanism to dynamo which can later be improved to be more thorough. Currently it allows tensors to be in tl.load but if it sees a tensor used anywhere else (including a tl.load), it bails out.

One question about the method: is `ast.NodeVisitor` the best thing to use here? Having to detect mutations with this is kinda pretty since you need to keep setting state at each transition.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Implement a simple mutation tracker for user defined triton kernels"


This PR adds a very simple mutation tracking mechanism to dynamo which can later be improved to be more thorough. Currently it allows tensors to be in tl.load but if it sees a tensor used anywhere else (including a tl.load), it bails out.

One question about the method: is `ast.NodeVisitor` the best thing to use here? Having to detect mutations with this is kinda pretty since you need to keep setting state at each transition.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge
